File: fixtures/coding/maintainability-mini-service-go/validation.go
package miniservice

import "strings"

var allowedPlans = map[string]struct{}{
	"free":       {},
	"pro":        {},
	"enterprise": {},
}

type NormalizedCustomerPayload struct {
	Name  string
	Email string
	Plan  string
	Tags  []string
}

// ValidateCustomerPayload validates the existing single-customer create payload.
func ValidateCustomerPayload(payload map[string]any) []string {
	errors := []string{}
	name, nameOK := payload["name"].(string)
	email, emailOK := payload["email"].(string)
	plan, planOK := payload["plan"].(string)
	if !planOK {
		plan = "free"
	}
	tags, tagsOK := payload["tags"].([]any)
	if _, exists := payload["tags"]; !exists {
		tagsOK = true
	}

	if !nameOK || strings.TrimSpace(name) == "" {
		errors = append(errors, "name is required")
	}
	if !emailOK || !isValidEmail(email) {
		errors = append(errors, "email must be valid")
	}
	if _, ok := allowedPlans[plan]; !ok {
		errors = append(errors, "plan is invalid")
	}
	if !tagsOK || !allStrings(tags) {
		errors = append(errors, "tags must be a list of strings")
	}

	return errors
}

// NormalizeCustomerPayload normalizes the existing single-customer create payload.
func NormalizeCustomerPayload(payload map[string]any) NormalizedCustomerPayload {
	plan, ok := payload["plan"].(string)
	if !ok {
		plan = "free"
	}
	return NormalizedCustomerPayload{
		Name:  strings.TrimSpace(asString(payload["name"])),
		Email: strings.ToLower(strings.TrimSpace(asString(payload["email"]))),
		Plan:  plan,
		Tags:  normalizeTags(payload["tags"]),
	}
}

func isValidEmail(email string) bool {
	at := strings.Index(email, "@")
	return at > 0 && strings.Contains(email[at+1:], ".")
}

func allStrings(values []any) bool {
	for _, value := range values {
		if _, ok := value.(string); !ok {
			return false
		}
	}
	return true
}

func normalizeTags(value any) []string {
	values, ok := value.([]any)
	if !ok {
		return []string{}
	}
	tags := make([]string, 0, len(values))
	for _, item := range values {
		tags = append(tags, strings.ToLower(strings.TrimSpace(asString(item))))
	}
	return tags
}

func asString(value any) string {
	text, _ := value.(string)
	return text
}

File: fixtures/coding/maintainability-mini-service-rust/src/validation.rs
use crate::models::{JsonMap, JsonValue};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct NormalizedCustomerPayload {
    pub name: String,
    pub email: String,
    pub plan: String,
    pub tags: Vec<String>,
}

/// Validates the existing single-customer create payload.
pub fn validate_customer_payload(payload: &JsonMap) -> Vec<String> {
    let mut errors = Vec::new();
    let name = string_field(payload, "name");
    let email = string_field(payload, "email");
    let plan = string_field(payload, "plan").unwrap_or("free");

    if name.is_none_or(|value| value.trim().is_empty()) {
        errors.push("name is required".to_string());
    }
    if email.is_none_or(|value| !is_valid_email(value)) {
        errors.push("email must be valid".to_string());
    }
    if !matches!(plan, "free" | "pro" | "enterprise") {
        errors.push("plan is invalid".to_string());
    }
    if !tags_are_strings(payload.get("tags")) {
        errors.push("tags must be a list of strings".to_string());
    }

    errors
}

/// Normalizes the existing single-customer create payload.
pub fn normalize_customer_payload(payload: &JsonMap) -> NormalizedCustomerPayload {
    NormalizedCustomerPayload {
        name: string_field(payload, "name")
            .unwrap_or_default()
            .trim()
            .to_string(),
        email: string_field(payload, "email")
            .unwrap_or_default()
            .trim()
            .to_lowercase(),
        plan: string_field(payload, "plan").unwrap_or("free").to_string(),
        tags: normalize_tags(payload.get("tags")),
    }
}

pub fn string_field<'a>(payload: &'a JsonMap, key: &str) -> Option<&'a str> {
    match payload.get(key) {
        Some(JsonValue::String(value)) => Some(value),
        _ => None,
    }
}

fn is_valid_email(email: &str) -> bool {
    let Some((local, domain)) = email.split_once('@') else {
        return false;
    };
    !local.is_empty() && domain.contains('.')
}

fn tags_are_strings(value: Option<&JsonValue>) -> bool {
    match value {
        None => true,
        Some(JsonValue::Array(values)) => values
            .iter()
            .all(|value| matches!(value, JsonValue::String(_))),
        _ => false,
    }
}

fn normalize_tags(value: Option<&JsonValue>) -> Vec<String> {
    let Some(JsonValue::Array(values)) = value else {
        return Vec::new();
    };
    values
        .iter()
        .filter_map(|value| match value {
            JsonValue::String(text) => Some(text.trim().to_lowercase()),
            _ => None,
        })
        .collect()
}
